### ISSUE TITLE
Implement support in Java for shared memory buffers

### ISSUE THREAD

**Issue by ctrueden** [apposed/appose#7]
I started implementing a `SharedMemory` class in Appose Java, which is a direct translation of Python's handy `multiprocessing.shared_memory.SharedMemory` class. It leans on JNA for the low-level system calls. But it is not yet fully working. This needs to be finished, so that we can share memory easily between processes!

I'm hoping the JNA-based approach will work directly. But other options include:
* Add support for shared memory buffers to the [LArray](https://github.com/xerial/larray) library—work already started by @mkitti [on a branch](https://github.com/mkitti/larray/compare/mkitti/named_mmap)
* Write our own small C library and ship it in the Appose Java JAR

**Comment by carlosuc3m** [comment]
HEllo Curtis,
I just finished implementing shared memory IPC communication on the JDLL side:
https://github.com/bioimage-io/JDLL/blob/python-appose/src/main/java/io/bioimage/modelrunner/tensor/SharedMemoryArray.java
I have used jna-platform which contains a ready to use JNA for kernel32 and posix methods (what cpython uses for [multiprocessing.shared_memory](https://docs.python.org/3/library/multiprocessing.shared_memory.html#module-multiprocessing.shared_memory)).
I had the urgency to implement it because when I tried to send relatively big images through the pipes it would take too much time. Changing to shm works great and has a huge performance improvement as the information sent by the pipes is way smaller now.
I did not change anything on the appose implementatio, I just carefully desigined the inputs that I sent to appose from JDLL and the outputs of the Python worker.

JDLL creates a script that is able to recreate the imglib2 array of the Java side:
```python
# Create the shared memory object pointint to the shm block of interest with the wanted size
input_06112023_151243_shm_a9b8d59c_0087_4394_a9dd_fb54e23235e6 = shared_memory.SharedMemory(name='b024e140-5780-4ec6-a45d-f51cfa2ce3e7', size=20545536)

# Recreate the data array with the data from the shm block
input_06112023_151243 = xr.DataArray(np.ndarray(5136384, dtype='float32', buffer=input_06112023_151243_shm_a9b8d59c_0087_4394_a9dd_fb54e23235e6.buf).reshape([1, 304, 512, 33]), dims=["b", "y", "x", "c"], name="output")
```

All the code above is then fed to the Python worker to be executed, as you can see the memory location, size and shape is hardcoded on the JAva side (done [here](https://github.com/bioimage-io/JDLL/blob/cdf5469f0405be6aabc30d7380ba0f4c1baf2a11/src/main/java/io/bioimage/modelrunner/runmode/RunMode.java#L295))


Then once the result is obtained, on the Python side, the shm block is created and the info to access that memory block is sent to java together with the data type, shape and whether is fortran order or not.

On the Python side:
```python
shm_out_list = []
def convertNpIntoDic(np_arr):
  shm = shared_memory.SharedMemory(create=True, size=np_arr.nbytes)
  aux_np_arr = np.ndarray((np_arr.size), dtype=np_arr.dtype, buffer=shm.buf)
  aux_np_arr[:] = np_arr.flatten()
  shm_out_list.append(shm)
  shm.unlink()
  return {"data": shm.name, "shape": np_arr.shape, "appose_data_type__06112023_151246": "np_arr", "is_fortran": np.isfortran(np_arr), "dtype": str(np_arr.dtype)}

def convertXrIntoDic(xr_arr):
  shm = shared_memory.SharedMemory(create=True, size=xr_arr.values.nbytes)
  aux_np_arr = np.ndarray((xr_arr.values.size), dtype=xr_arr.values.dtype, buffer=shm.buf)
  aux_np_arr[:] = xr_arr.values.flatten()
  shm_out_list.append(shm)
  shm.unlink()
  return {"data": shm.name, "shape": xr_arr.shape, "axes": "".join(xr_arr.dims),"name": xr_arr.name, "appose_data_type__06112023_151246": "tensor", "is_fortran": np.isfortran(xr_arr.values), "dtype": str(xr_arr.values.dtype)}
```

The dictionary returned is then used by the following Java method to recreate a JDLL tensor or ImgLib array:
https://github.com/bioimage-io/JDLL/blob/cdf5469f0405be6aabc30d7380ba0f4c1baf2a11/src/main/java/io/bioimage/modelrunner/tensor/SharedMemoryArray.java#L422

Performance is improved greatly (at least on my old laptop) because the stdout does not need to flush millions of numbers and the result looks like the following:
https://twitter.com/carlosg91018370/status/1721526937686868443

A similar logic dould be implemented in Appose to support sending more high level objects such as numpy arrays or imglib2 images.

Regards,
Carlos

**Comment by ctrueden** [comment]
This issue has been addressed by apposed/appose-java#5. Big thanks to @tpietzsch and @carlosuc3m for their efforts. :beers:

**Comment by ctrueden** [comment]
And for the record, the Python implementation maintains feature parity with Java thanks to apposed/appose-python#1.
